### PR TITLE
SK-717: Repair faulty DIDs

### PR DESCRIPTION
**Pull request by lukeg90**
SK-717

### CODE REVIEW

File: src/views/IdentityOverview/IdentityOverview.test.tsx [hunk 1]
```diff
@@ -25,6 +31,8 @@ const identity =
 const fullDidIdentity =
   identitiesMock['4sm9oDiYFe22D7Ck2aBy5Y2gzxi2HhmGML98W9ZD2qmsqKCr'];
 
+const legacyIdentity = legacyIdentityMock;
```
Comment (arty-name) [inline src/views/IdentityOverview/IdentityOverview.test.tsx:34]: Here as well

File: src/views/IdentityOverview/IdentityOverview.tsx [hunk 2]
```diff
@@ -45,6 +45,22 @@ export function IdentityOverview({ identity }: Props): JSX.Element | null {
...
     credentials && credentials.some(({ isDownloaded }) => !isDownloaded);
 
+  const [hasLegacyDid, setHasLegacyDid] = useState(false);
+
+  useEffect(() => {
+    (async () => {
+      if (await needLegacyDidCrypto(identity.did)) {
+        setHasLegacyDid(true);
+      } else {
+        setHasLegacyDid(false);
```
Comment (arty-name) [inline src/views/IdentityOverview/IdentityOverview.tsx:55]: It is already false by default, right?
```suggestion
```

File: src/views/IdentityOverview/IdentityOverview.tsx [hunk 3]
```diff
@@ -45,6 +45,22 @@ export function IdentityOverview({ identity }: Props): JSX.Element | null {
...
+      }
+    })();
+  }, [identity, hasLegacyDid]);
```
Comment (arty-name) [inline src/views/IdentityOverview/IdentityOverview.tsx:58]: Does it really depend on hasLegacyDid?
```suggestion
  }, [identity]);
```

File: src/views/IdentityOverview/IdentityOverview.tsx [hunk 4]
```diff
@@ -126,6 +142,17 @@ export function IdentityOverview({ identity }: Props): JSX.Element | null {
         </Link>
       )}
 
+      {repairDid && (
+        <Link
+          to={generatePath(paths.identity.did.repair, {
+            address: identity.address,
+          })}
```
Comment (arty-name) [inline src/views/IdentityOverview/IdentityOverview.tsx:149]: We already have address in scope
```suggestion
          to={generatePath(paths.identity.did.repair, { address })}
```

File: src/views/IdentityOverview/IdentityOverview.tsx (hunk 2 above)
Comment (lukeg90) [reply src/views/IdentityOverview/IdentityOverview.tsx:55]: I initially had the same thought, but leaving out this line causes the 'true' state to persist when the identity is changed with the arrow buttons.

File: src/utilities/didRepair/didRepair.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+  const existingKeyId = DidUtils.parseDidUrl(existingKey.id).fragment;
+
+  const { encryptionKey: newKeyId } = deriveDidKeys(identity);
+
+  const { api } = await BlockchainApiConnection.getConnectionOrConnect();
+
+  const batchExtrinsic = api.tx.utility.batchAll([
+    await DidChain.getRemoveKeyExtrinsic(keyAgreement, existingKeyId),
+    await DidChain.getAddKeyExtrinsic(keyAgreement, newKeyId),
```
[the remark this answers is not in the document]
Comment (lukeg90) [reply src/utilities/didRepair/didRepair.ts:56]: Let's go with `existingKey` and `newKey`?

File: src/utilities/identities/identities.ts [hunk 6]
```diff
@@ -146,7 +146,17 @@ function extractSecretKey(keypair: KeyringPair) {
...
+export function deriveDidKeys(
+  identityKeypair: KeyringPair,
+  legacy?: boolean,
+): {
+  authenticationKey: KeyringPair;
+  encryptionKey: {
+    type: string;
+    publicKey: Uint8Array;
+    secretKey: Uint8Array;
+  };
```
[the remark this answers is not in the document]
Comment (lukeg90) [reply src/utilities/identities/identities.ts:158]: Are you sure this is the right type? It doesn't contain a `secretKey` property

File: src/components/LegacyDids/LegacyDids.tsx [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+export function LegacyDids(): JSX.Element | null {
+  const t = browser.i18n.getMessage;
+
+  const open = useBooleanState();
+
+  const legacyDidIdentities = useLegacyDidIdentities();
+
+  useEffect(() => {
+    open.set(Object.values(legacyDidIdentities).length > 0);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (arty-name) [inline src/components/LegacyDids/LegacyDids.tsx:23]: Let’s have a better look at why this line would be needed.

File: src/components/LegacyDids/LegacyDids.tsx [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  }, [legacyDidIdentities]);
+
+  return (
+    <Modal open={open.current} className={styles.overlay}>
```
Comment (arty-name) [inline src/components/LegacyDids/LegacyDids.tsx:27]: I think this way we will not have a bunch of invisible DOM elements all the time, and the performance would be better.

```suggestion
  if (!open.current) {
    return null;
  }
  
  return (
    <Modal open className={styles.overlay}>
```

File: src/utilities/didRepair/didRepair.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (arty-name) [reply src/utilities/didRepair/didRepair.ts]: The current state with `existingKeyId` and `encryptionKey` matches what the APIs expect and looks good.

File: src/utilities/identities/identities.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (arty-name) [reply src/utilities/identities/identities.ts:158]: Yeah, I was wrong here

**Review by arty-name — APPROVED** [review]

**Pull request merged**